p4-16/spec/P4-16-spec.mdk

```diff
@@ -4762,6 +4768,36 @@ H h2 = { f2=5, ... };   // initialize h2 with a header that is valid, field f1 0
 H h3 = { ... };  // initialize h3 with a header that is valid, field f1 0, field f2 0
 ~ End P4Example
 
+# Compile-time size determination { #sec-minsizeinbits }
+
+The method calls `minSizeInBits` and `minSizeInBytes` can be applied
+to most expressions.  Each of these method calls evaluate to
+compile-time known values that return the minimum size in bits
+required to store the expression.
+
+The method `minSizeInBytes` always returns the result the result of
+`minSizeInBits` rounded up to the next whole number of bytes.  not a
+multiple of 8 bits long.  In other words, for any expression `e`,
+`e.minSizeInBytes()` is equal to `(e.minSizeInBits() + 7) >> 3`.
+
+The definition of `e.minSizeInBits` is given recursively on the type of `e` as
+follows:
+
+* if the type of `e` is `varbits` the result is 0
```

Comment:
Do we want this semantics? And is it consistent with what we do for header unions and stacks below?

Reply:
I can see the consistency with the definition of `minSizeInBits()` for the headers above (where we state that varbits are counted as 0).

However, for the practical purposes, it would be a lot more useful to have at least **some way** to return `N` for a `varbit<N>`.

One way to do that, might be to state that `e.minSizeInBits()` is really the minimum number of bits that should be allocated to store ANY value of the corresponding type. This, by the way, is consistent with what we do with header stacks that can also be only partially filled. This also works well with the idea that we allocate all resources statically and as a result there are no exceptions in the controls.